dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ZipGitHubFileTree.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2024 OICR and UCSC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.helpers;
+
+import io.dockstore.common.Utilities;
+import io.dockstore.webservice.CustomWebApplicationException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.apache.commons.compress.utils.SeekableInMemoryByteChannel;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstracts the file tree corresponding to a reference to a GitHub repository.
+ *
+ * Upon initialization, this implementation retrieves a Zipball of the ref/repo
+ * file tree and constructs a path-to-ZipArchiveEntry map corresponding to normal
+ * (non-symlink) files.  Sebsequently, on demand, the map is used to extract file
+ * contents and a list of paths.
+ *
+ * The zip file and path-to-entry map do not include submodule files or paths
+ * that contain a symlink component.  If `readFile` can't find a path in the map,
+ * we attempt to read the file via `GitHubSourceCodeRepo.readFile`, which
+ * supports symlinks and submodules.
+ */
+public class ZipGitHubFileTree implements FileTree {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ZipGitHubFileTree.class);
+
+    private final GitHubSourceCodeRepo gitHubSourceCodeRepo;
+    private final String repository;
+    private final String ref;
+    private final ZipFile zipFile;
+    private final Map<String, ZipArchiveEntry> pathToEntry;
+
+    public ZipGitHubFileTree(GitHubSourceCodeRepo gitHubSourceCodeRepo, String repository, String ref) {
+        this.gitHubSourceCodeRepo = gitHubSourceCodeRepo;
+        this.repository = repository;
+        this.ref = ref;
+        // Read the Zip contents and create an in-memory ZipFile.
+        byte[] zipBytes = gitHubSourceCodeRepo.readZip(repository, ref);
+        LOG.info("downloaded Zip of GitHub repository: %n bytes".formatted(zipBytes.length));
+        SeekableByteChannel zipChannel = new SeekableInMemoryByteChannel(zipBytes);
```

Comment:
May need to close this or use try with resources

Reply:
Nope, checked out the implementation, and as expected, it doesn't allocate any non-heap/OS resources:
https://github.com/apache/commons-compress/blob/acaf5642b1e16d872870a4dc4618e46ccd349493/src/main/java/org/apache/commons/compress/utils/SeekableInMemoryByteChannel.java
The channel needs to be open for the lifespan of the ZipGitHubFileTree, after which it'll become unreferenced and garbage collected.

Reply:
Pending vfs discussion anyway